File: as/app/src/main/java/com/trik/gamepad/StartReadMjpegAsync.java
// based on http://stackoverflow.com/questions/10550139/android-ics-and-mjpeg-using-asynctask

package com.trik.gamepad;

import android.os.AsyncTask;
import android.util.Log;

import com.demo.mjpeg.MjpegInputStream;
import com.demo.mjpeg.MjpegView;

import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;

import java.io.IOException;
import java.net.URI;

class StartReadMjpegAsync extends AsyncTask<URI, Void, MjpegInputStream> {
    private final MjpegView mv;

    StartReadMjpegAsync(MjpegView mv) {
        this.mv = mv;
    }

    @Override
    protected MjpegInputStream doInBackground(URI... uris) {
        HttpResponse res;
        URI uri = uris[0];
        if (uri != null) {
            try {
                DefaultHttpClient httpclient = new DefaultHttpClient();
                res = httpclient.execute(new HttpGet(uri));
                return new MjpegInputStream(res.getEntity().getContent());
            } catch (IOException e) {
                Log.e(this.getClass().getSimpleName(), Log.getStackTraceString(e));

            }
        }
        return null;
    }

    @Override
    protected void onPostExecute(MjpegInputStream result) {
        mv.setSource(result);
        mv.startPlayback();
    }
}

File: as/app/src/main/kotlin/com/trik/gamepad/StartReadMjpegAsync.kt
// based on http://stackoverflow.com/questions/10550139/android-ics-and-mjpeg-using-asynctask

package com.trik.gamepad

import android.os.AsyncTask
import com.demo.mjpeg.MjpegInputStream
import com.demo.mjpeg.MjpegView
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URI


internal class StartReadMjpegAsync(private val mv: MjpegView) : AsyncTask<URI, Void, MjpegInputStream>() {

    override fun doInBackground(vararg uris: URI): MjpegInputStream? {
        val uri = uris[0]
            try {
                val u = uri.toURL()
                val c = u.openConnection() as HttpURLConnection
                c.connectTimeout = 5000
                c.readTimeout = 5000
                return MjpegInputStream(c.inputStream)

            } catch (e: IOException) {
                e.printStackTrace()
            }
        return null
    }

    override fun onPostExecute(result: MjpegInputStream) {
        mv.setSource(result)
        mv.startPlayback()
    }
}
